client/lib/html-toolbar/index.js

```diff
@@ -0,0 +1,229 @@
+		return;
+	} else if ( /(?:edge|msie |trident.+?; rv:)/i.test( userAgent ) ) {
+		// execCommand( 'insertText' ), needed to preserve the undo stack, does not exist in IE11,
+		// and Edge has issues positioning the cursor after insertion.
+		textarea.value = newContent;
+		setCursorPosition( textarea, selectionEnd, newContent.length - value.length );
+		onInsert( newContent );
+		return;
+	}
+	textarea.focus();
+	document.execCommand( 'insertText', false, content );
+};
+
+/**
+ * Convert an object of attributes into a string.
+ *
+ * @param {Object} attributes The object of attributes.
+ * @returns {String} The string of attributes.
+ */
+export const attributesToString = ( attributes = {} ) =>
+	reduce(
+		attributes,
+		( attributesString, attributeValue, attributeName ) =>
+			attributeValue
+				? attributesString + ` ${ attributeName }="${ attributeValue }"`
+				: attributesString,
+		''
+	);
+
+/**
+ * Creates an open tag with the given attributes and options.
+ * E.g. `<div class="foo">`
+ *
+ * @param {Object} tag The tag to open.
+ * @returns {String} The open tag.
+ */
+const openHtmlTag = ( { name, attributes = {}, options = {} } ) =>
+	( options.paragraph ? '\n' : '' ) +
+	( options.indent ? '\t' : '' ) +
+	`<${ name }${ attributesToString( attributes ) }>` +
+	( options.paragraph ? '\n' : '' );
+
+/**
+ * Creates a closed tag with the given options.
+ * E.g. `</div>`
+ *
+ * @param {Object} tag The tag to close.
+ * @returns {String} The closed tag.
+ */
+export const closeHtmlTag = ( { name, options = {} } ) =>
+	`</${ name }>` + ( options.newLineAfter ? '\n' : '' ) + ( options.paragraph ? '\n\n' : '' );
+
+/**
+ * Insert an open tag into the textarea.
+ *
+ * @param {Element} textarea The editor textarea.
+ * @param {Object} tag The tag to insert.
+ * @param {Function} onInsert A function to call after the insertion.
+ */
+export const insertHtmlTagOpen = ( textarea, tag, onInsert ) => {
+	const { before, after } = splitSelectedContent( textarea );
+	insertContent( textarea, before, openHtmlTag( tag ), after, onInsert );
+};
+
+/**
+ * Insert a closed tag into the textarea.
+ *
+ * @param {Element} textarea The editor textarea.
+ * @param {Object} tag The tag to insert.
+ * @param {Function} onInsert A function to call after the insertion.
+ */
+export const insertHtmlTagClose = ( textarea, tag, onInsert ) => {
+	const { before, after } = splitSelectedContent( textarea );
+	insertContent( textarea, before, closeHtmlTag( tag ), after, onInsert );
+};
+
+/**
+ * Insert, open, and close, a tag into the textarea.
+ *
+ * @param {Element} textarea The editor textarea.
+ * @param {Object} tag The tag to insert.
+ * @param {Function} onInsert A function to call after the insertion.
+ */
+export const insertHtmlTagOpenClose = ( textarea, tag, onInsert ) => {
+	const { before, inner, after } = splitSelectedContent( textarea );
+	insertContent(
+		textarea,
+		before,
+		openHtmlTag( tag ) + inner + closeHtmlTag( tag ),
```

Comment:
with as sensitive as HTML can be and as easy as it is to break it we could also consider using the browser's built-in DOM methods here

```js
const l = m => console.log( m ) || m;
console.clear();

const makeElement = ( name, attributes, content ) => {
  const e = document.createElement( name )
  
  if ( attributes ) {
    Object
      .keys( attributes )
      .forEach( key => e.setAttribute( key, attributes[ key ] ) )
  }
  
  if ( content ) {
    e.innerHTML = content
  }
  
  return e
}

const makeHtml = ( ...args ) => makeElement( ...args ).outerHTML

// there's a better way…
const makeHtmlPair = ( name, attributes ) => {
  // the HTML comment _cannot_ exist inside a tag
  const e = makeElement( name, attributes, '<!---->' )
  
  return e.outerHTML.split( '<!---->' ) 
}
```

http://jsbin.com/getuhemule/1/edit?js,console

Reply:
note for example that this takes HTML rules into play. I tried creating a pair for a `br` instead of a `p` and it only gave me one string back - because `br` _can't_ be split even though a simpler text-generation function can make it seem that way